### PR TITLE
Resolves #504: Introduce mocked locality API

### LINKED ISSUE #504: Mocked locality API
In discussions on PR #454, this came up: 

It would be a nice feature to have a record layer API that abstracts the implementation of the Locality API and, thus, allow you to plug your own implementation in.

### PR DESCRIPTION
**Pull request by amouehsan**
This change adds an interface for the locality API and implements one class that uses FDB's
LocalityUtil and one that fakes locality API by creating random boundaries. The interface is hooked
to FDBDatabase/FDBDatabaseFactory with a default that uses FDB LocalityUtil.

### CODE REVIEW

**Comment by amouehsan** [comment]
Rebased against master and resolved the conflicts.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java [hunk 1]
```diff
@@ -2286,30 +2287,38 @@ public void formerIndexes() throws Exception {
 
     @Test
     public void boundaryPrimaryKeys() throws Exception {
```
Comment (nschiefer) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: This PR should probably revert the change from `assertTrue` -> `assumeTrue` in #567.

Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: Good point. Thanks for pointing out.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    public CloseableAsyncIterator<byte[]> getBoundaryKeys(@Nonnull Transaction tr, @Nonnull byte[] begin, @Nonnull byte[] end) {
+        // This mocked locality API does not need tr.
+        return new MockedBoundaryIterator(keyRanges, ranges, begin, end);
+    }
+
+    static class MockedBoundaryIterator implements CloseableAsyncIterator<byte[]> {
+        private List<byte[]> ranges;
+        int lastBeginIndex;
+
+        MockedBoundaryIterator(@Nonnull List<KeyRange> keyRanges, @Nonnull List<Integer> ranges, @Nonnull byte[] begin, @Nonnull byte[] end) {
```
Comment (nblintao) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java]: There should be JavaDoc explaining `keyRanges` and `ranges`

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabase.java [hunk 3]
```diff
@@ -924,8 +936,8 @@ private void logOrThrowBlockingInAsync(@Nonnull BlockingInAsyncDetection behavio
     public CompletableFuture<List<Tuple>> computeBoundaryKeys(@Nonnull FDBTransactionContext context, Tuple prefix,
```
[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabase.java]: Remove it in favor of `FDBRecordStore.computeBoundaryKeys` or whatever that method is called?

I think that sounds reasonable enough. I think that most needs should be served by that other method anyway.

[the remark this answers is not in the document]
Comment (amouehsan) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabase.java]: It's a stable public method. Would it be fine to just remove it? How about I deprecate it for now?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseFactory.java [hunk 4]
```diff
@@ -494,4 +497,22 @@ public synchronized FDBDatabase getDatabase(@Nullable String clusterFile) {
...
         return getDatabase(null);
     }
+
+    /**
+     * Get the locality provider that is used to discover the server location of the keys.
+     * @return the installed locality provider
+     */
+    @Nonnull
+    public FDBLocalityProvider getLocalityProvider() {
+        return this.localityProvider;
```
[the remark this answers is not in the document]
Comment (amouehsan) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseFactory.java]: sure, done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBLocalityUtil.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.LocalityUtil;
+import com.apple.foundationdb.Transaction;
+import com.apple.foundationdb.async.CloseableAsyncIterator;
+
+import javax.annotation.Nonnull;
+
+/**
+ * An implementation of the {@link FDBLocalityProvider} interface that uses foundationDB's {@link LocalityUtil} API to
```
[the remark this answers is not in the document]
Comment (amouehsan) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBLocalityUtil.java]: This is what InteliJ suggested. Do you still want this to be changed? I'm ok either way.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabase.java]: It's marked as `INTERNAL`, so by our stability rules, it can be removed at any time. I don't think there's any harm, though, in deprecating it first.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java [hunk 6]
```diff
@@ -2287,30 +2287,38 @@ public void formerIndexes() throws Exception {
...
+        final FDBDatabaseFactory factory = FDBDatabaseFactory.instance();
+        factory.clear();
```
[the remark this answers is not in the document]
Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: I don't recall why it was needed. Test works without it, so dropping it. Thanks for catching this.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java [hunk 7]
```diff
@@ -2287,30 +2287,38 @@ public void formerIndexes() throws Exception {
...
+        try (FDBRecordContext context = database.openContext()) {
             openSimpleRecordStore(context, TEST_SPLIT_HOOK);
             recordStore.markIndexWriteOnly(indexName).get();
             commit(context);
         }
 
+        ArrayList<byte[]> keys = new ArrayList<>();
         String bigOlString = Strings.repeat("x", SplitHelper.SPLIT_RECORD_SIZE + 2);
         for (int i = -25; i < 25; i++) {
             // Sparsify the primary keys so the records can be saved across boundaries.
```
[the remark this answers is not in the document]
Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: sure, didn't change it because I didn't see any harm in doing it. changed now.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A mocked implementation of the {@link FDBLocalityProvider} interface.
+ */
+public class MockedLocalityUtil implements FDBLocalityProvider {
```
[the remark this answers is not in the document]
Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java]: Good point. Added `testMockedLocalityUtil` let me know if you have tests suggestions.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        private int rangeIndex;
+
+        public KeyRange(@Nonnull byte[] key, @Nonnull Tuple tuple, int rangeIndex) {
+            this.key = key;
+            this.tuple = tuple;
+            this.rangeIndex = rangeIndex;
+        }
+    }
+
+    private static final MockedLocalityUtil INSTANCE = new MockedLocalityUtil();
```
[the remark this answers is not in the document]
Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java]: My assumption was in the beginning of every test, init should be called only once and tests won't race each other. Is this not true?

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    /**
+     * Initialize the {@code MockedLocalityUtil}'s singleton using a random number of ranges.
+     * @param keys a sorted list of keys
+     * @see #init(List, int)
+     */
+    @Nonnull
+    public static void init(@Nonnull List<byte[]> keys) {
+        init(keys, new Random().nextInt(keys.size()) + 1);
```
[the remark this answers is not in the document]
Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java]: It's not used. I added it as a convenience API and thought the reasoning behind the range is obvious. I can drop it.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java (hunk 2 above)
Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java]: Sure, I can add it, but the scope of this and the two fields is not `public`.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (amouehsan) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabase.java]: oh I missed that. Lets just remove it then.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBLocalityUtil.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (nblintao) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBLocalityUtil.java]: Never mind. I was suggesting because I thought it had a same name with the new class.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (nblintao) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java]: If two tests are running at the same time, one may use the `INSTANCE` after the other re`init` it. I guess we are not running tests concurrently in out CI today but maybe we want to be protected against it? @alecgrieser

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java [hunk 11]
```diff
@@ -2451,4 +2451,43 @@ private void checkSplitIndexBuildRange(int minSplit, int maxSplit,
...
+            assertEquals(rangeCount, getRangeCount(MockedLocalityUtil.instance().getBoundaryKeys(context.ensureActive(), keys.get(0), keys.get(keys.size() - 1))));
+
+            MockedLocalityUtil.init(keys, keys.size());
+            assertEquals(keys.size(), getRangeCount(MockedLocalityUtil.instance().getBoundaryKeys(context.ensureActive(), keys.get(0), keys.get(keys.size() - 1))));
+
+            commit(context);
+        }
+    }
+
+    private int getRangeCount(@Nonnull CloseableAsyncIterator<byte[]> iterator) {
```
Comment (nblintao) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: I guess you can use Guavva's `Iterators.size(iterator)`

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java [hunk 12]
```diff
@@ -2451,4 +2451,43 @@ private void checkSplitIndexBuildRange(int minSplit, int maxSplit,
...
+            assertEquals(0, getRangeCount(MockedLocalityUtil.instance().getBoundaryKeys(context.ensureActive(), keys.get(0), keys.get(keys.size() - 1))));
+
+            MockedLocalityUtil.init(keys, 1);
+            assertEquals(1, getRangeCount(MockedLocalityUtil.instance().getBoundaryKeys(context.ensureActive(), keys.get(0), keys.get(keys.size() - 1))));
+
+            int rangeCount = new Random().nextInt(keys.size() + 1);
+            MockedLocalityUtil.init(keys, rangeCount);
+            assertEquals(rangeCount, getRangeCount(MockedLocalityUtil.instance().getBoundaryKeys(context.ensureActive(), keys.get(0), keys.get(keys.size() - 1))));
+
+            MockedLocalityUtil.init(keys, keys.size());
```
Comment (nblintao) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: Is it safe to have a range count greater than `keys.size()`?

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java [hunk 13]
```diff
@@ -2451,4 +2451,43 @@ private void checkSplitIndexBuildRange(int minSplit, int maxSplit,
...
+
+        try (FDBRecordContext context = database.openContext()) {
+            openSimpleRecordStore(context, TEST_SPLIT_HOOK);
+            ArrayList<byte[]> keys = new ArrayList<>();
+            for (int i = 0; i < 50; i++) {
+                keys.add(recordStore.recordsSubspace().pack(i));
+            }
+
+            MockedLocalityUtil.init(keys, 0);
+            assertEquals(0, getRangeCount(MockedLocalityUtil.instance().getBoundaryKeys(context.ensureActive(), keys.get(0), keys.get(keys.size() - 1))));
```
Comment (nblintao) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: Maybe this can be pulled out.
```java
testRangeCount(0);
testRangeCount(1);
...
private testRangeCount(int rangeCount) {
    MockedLocalityUtil.init(keys, rangeCount);
    ... boudnaryKeys = MockedLocalityUtil.instance()
        .getBoundaryKeys(context.ensureActive(), keys.get(0), keys.get(keys.size() - 1));
    assertEquals(rangeCount, getRangeCount(boudnaryKeys));
}
```

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java [hunk 14]
```diff
@@ -2354,8 +2362,13 @@ public void boundaryPrimaryKeys() throws Exception {
 
     @Test
     public void noBoundaryPrimaryKeys() throws Exception {
+        final FDBDatabaseFactory factory = FDBDatabaseFactory.instance();
+        factory.clear();
```
[the remark this answers is not in the document]
Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: sure, done.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java (hunk 11 above)
Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: Good point. Done.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java (hunk 13 above)
Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: good idea, done.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java (hunk 12 above)
Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: Not really. Let me disallow it.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java]: Right, I see that. But if no tests are running in parallel, probably there's not a need for it just yet?

[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java]: At the moment, I think there are a fair number of barriers to parallelized tests, including the fact that many tests write to the same place in the FDB cluster, so they could very well hit each other in the face. That being said, I don't think we should necessarily make the problem worse, if it's an easy fix. (I haven't looked into this case in particular at all.)

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java [hunk 15]
```diff
@@ -2354,8 +2362,12 @@ public void boundaryPrimaryKeys() throws Exception {
 
     @Test
     public void noBoundaryPrimaryKeys() throws Exception {
+        final FDBDatabaseFactory factory = FDBDatabaseFactory.instance();
+        factory.setLocalityProvider(MockedLocalityUtil.instance());
+        FDBDatabase database = FDBDatabaseFactory.instance().getDatabase();
```
Comment (scgray) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: I think you *probably* have a problem here.  The `FDBDatabaseFactory` maintains a cache of the database instances, so if another test has already created the `FDBDatabase` then you will be using the `FDBLocalityProvider` that was previously installed.  Or, if you do install the `MockedLocalityUtil` provider, the next test that runs and needs a locality provider will end up using the one you just installed.

It would probably make sense to add a `@Before` method that does: `FDBDatabaseFactory.instance().clear()`. This will at least remove the cached database instances, but it won't take care of the locality API that you've installed.  I guess that could be handled in the test cases themselves for now.

 Or, optional, in the JUnit4 world, we could have had a:

    @Rule
    FDBDatabaseFactoryRule databaseFactory = new FDBDatabaseFactoryRule();
    ...
    FDBDatabase database = databaseFactory.getDatabase();

that would automatically reset the cached databases before each test...it appears that the way to do this in JUnit5 is with `@Extension`, but that might be a bit much for this PR (although, it would generally be very useful).   I think I'll open a separate issue for this.

Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: Yeah, that's right! In fact, I used to call `clear()` but later when @nblintao asked about the reason, I didn't recall and dropped it as the tests passed without it. Now I recall, I had hit the exact same problem. I'll fix the immediate problem with the easy way, and open an issue (I've never used the `@Extension`  before - so I'll defer it but will get back to it soon). Thanks for the review.

Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: It should be fixed now. I think you opened an issue for this already (#591). thanks.

**Comment by nblintao** [comment]
@amouehsan Just noticed that the release notes are not changed.

**Comment by amouehsan** [comment]
> @amouehsan Just noticed that the release notes are not changed.

Good point. Added now.

**Comment by scgray** [comment]
@amouehsan the proto3 build failed with:

```
com.apple.foundationdb.record.provider.foundationdb.OnlineIndexerTest > fiftySplitElements() FAILED
14:39:36     com.apple.foundationdb.record.provider.foundationdb.FDBExceptions$FDBStoreRetriableException: Request for future version
14:39:36         at com.apple.foundationdb.record.provider.foundationdb.FDBExceptions.wrapException(FDBExceptions.java:153)
14:39:36         at com.apple.foundationdb.record.provider.foundationdb.FDBDatabase.lambda$new$0(FDBDatabase.java:156)
14:39:36         at com.apple.foundationdb.record.provider.foundationdb.FDBDatabase.asyncToSync(FDBDatabase.java:790)
14:39:36         at com.apple.foundationdb.record.provider.foundationdb.FDBRecordContext.asyncToSync(FDBRecordContext.java:434)
14:39:36         at com.apple.foundationdb.record.provider.foundationdb.FDBRecordStoreBase$BaseBuilder.createOrOpen(FDBRecordStoreBase.java:1883)
14:39:36         at com.apple.foundationdb.record.provider.foundationdb.OnlineIndexerTest.openContext(OnlineIndexerTest.java:204)
14:39:36         at com.apple.foundationdb.record.provider.foundationdb.OnlineIndexerTest.openContext(OnlineIndexerTest.java:214)
14:39:36         at com.apple.foundationdb.record.provider.foundationdb.OnlineIndexerTest.lambda$valueRebuild$10(OnlineIndexerTest.java:447)
14:39:36         at com.apple.foundationdb.record.provider.foundationdb.OnlineIndexerTest.singleRebuild(OnlineIndexerTest.java:306)
14:39:36         at com.apple.foundationdb.record.provider.foundationdb.OnlineIndexerTest.valueRebuild(OnlineIndexerTest.java:510)
14:39:36         at com.apple.foundationdb.record.provider.foundationdb.OnlineIndexerTest.valueRebuild(OnlineIndexerTest.java:523)
14:39:36         at com.apple.foundationdb.record.provider.foundationdb.OnlineIndexerTest.fiftySplitElements(OnlineIndexerTest.java:969)
```

**Comment by alecgrieser** [comment]
That's almost certainly an environment thing. In particular, this error comes from the FDB server not being up to date with all of the most recent changes, which can happen, if, say, the disk were slow. I think we've seen that go up recently, for whatever reason.

**Comment by amouehsan** [comment]
Then, lets see if it fails again. 

test this please

**Comment by amouehsan** [comment]
Failed again with a different error: 
`org.opentest4j.AssertionFailedError: expected: <5> but was: <4>`

I'll take a look.

**Comment by amouehsan** [comment]
The latest failure appeared to be a test bug that didn't handle a corner case well. It should be fixed now.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java [hunk 16]
```diff
@@ -2358,6 +2366,8 @@ public void boundaryPrimaryKeys() throws Exception {
         checkSplitIndexBuildRange(boundaryPrimaryKeysSize / 2, Integer.MAX_VALUE, oneRangePerSplit, indexer); // to test that integer overflow isn't a problem
 
         indexer.close();
+        factory.unsetLocalityProvider();
```
Comment (scgray) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: This will still have problems if the test fails -- you'll never reset the provider.  How about something like:

    @Test   
    public void testSomething() {
        runLocalityTest( () -> testSomethingImpl() );
    }

    private void testSomethingImpl() {
    ...
    }

    private void runLocalityTest(Runnable test) {
        final FDBLocalityProvider origProvider = factory.getLocalityProvider();
        try {
            test.run();
        } finally {
            factory.setLocalityProvider(origProvider);
        }
    }

Comment (amouehsan) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreIndexTest.java]: Yeah you are right. Did what you suggested.
